This patch is the start of storing/using checksum information from
extended attribute values.  The rsync code only reads the values
at the moment.  There is also a perl script that can create them.

To use this patch, run these commands for a successful build:

    patch -p1 <patches/checksum-xattr.diff
    ./configure                               (optional if already run)
    make

--- old/flist.c
+++ new/flist.c
@@ -1193,7 +1193,8 @@ struct file_struct *make_file(const char
 	}
 #endif
 
-	if (always_checksum && am_sender && S_ISREG(st.st_mode))
+	if (always_checksum && am_sender && S_ISREG(st.st_mode)
+	 && !get_sum_xattr(thisname, &st, tmp_sum))
 		file_checksum(thisname, tmp_sum, st.st_size);
 
 	F_PATHNAME(file) = pathname;
--- old/generator.c
+++ new/generator.c
@@ -627,7 +627,8 @@ int unchanged_file(char *fn, struct file
 	   of the file time to determine whether to sync */
 	if (always_checksum > 0 && S_ISREG(st->st_mode)) {
 		char sum[MAX_DIGEST_LEN];
-		file_checksum(fn, sum, st->st_size);
+		if (!get_sum_xattr(fn, st, sum))
+			file_checksum(fn, sum, st->st_size);
 		return memcmp(sum, F_SUM(file), checksum_len) == 0;
 	}
 
--- old/support/xsums
+++ new/support/xsums
@@ -0,0 +1,118 @@
+#!/usr/bin/perl -w
+use strict;
+
+use Getopt::Long;
+use Cwd qw(abs_path cwd);
+use Digest::MD4;
+use Digest::MD5;
+use File::ExtAttr ':all';
+
+our($recurse_opt, $help_opt);
+our $verbosity = 0;
+
+&Getopt::Long::Configure('bundling');
+&usage if !&GetOptions(
+    'recurse|r' => \$recurse_opt,
+    'verbose|v+' => \$verbosity,
+    'help|h' => \$help_opt,
+) || $help_opt;
+
+my $start_dir = cwd();
+
+my @dirs = @ARGV;
+@dirs = '.' unless @dirs;
+foreach (@dirs) {
+    $_ = abs_path($_);
+}
+
+$| = 1;
+
+my $md4 = Digest::MD4->new;
+my $md5 = Digest::MD5->new;
+
+while (@dirs) {
+    my $dir = shift @dirs;
+
+    if (!chdir($dir)) {
+	warn "Unable to chdir to $dir: $!\n";
+	next;
+    }
+    if (!opendir(DP, '.')) {
+	warn "Unable to opendir $dir: $!\n";
+	next;
+    }
+
+    if ($verbosity) {
+	my $reldir = $dir;
+	$reldir =~ s#^$start_dir(/|$)# $1 ? '' : '.' #eo;
+	print "scanning $reldir\n";
+    }
+
+    my @subdirs;
+    while (defined(my $fn = readdir(DP))) {
+	next if $fn =~ /^\.\.?$/ || -l $fn;
+	if (-d _) {
+	    push(@subdirs, "$dir/$fn");
+	    next;
+	}
+	next unless -f _;
+
+	my($size,$mtime) = (stat(_))[7,9];
+
+	my $sum4 = getfattr($fn, 'rsync.%md4');
+	my $sum5 = getfattr($fn, 'rsync.%md5');
+
+	foreach ($sum4, $sum5) {
+	    if (defined $_) {
+		if (length($_) == 24) {
+		    my($sz,$mt,$sum) = unpack('V2a16', $_);
+		    if ($sz != ($size & 0xFFFFFFFF)
+		     || $mt != ($mtime & 0xFFFFFFFF)) {
+			$_ = undef;
+		    } else {
+			$_ = $sum;
+		    }
+		} else {
+		    $_ = undef;
+		}
+	    }
+	}
+	if (!defined($sum4) || !defined($sum5)) {
+	    if (!open(IN, $fn)) {
+		print STDERR "Unable to read $fn: $!\n";
+		next;
+	    }
+
+	    while (sysread(IN, $_, 64*1024)) {
+		$md4->add($_);
+		$md5->add($_);
+	    }
+	    close IN;
+
+	    $sum4 = $md4->digest;
+	    $sum5 = $md5->digest;
+	    print " $fn\n" if $verbosity > 1;
+
+	    my $szmt = pack('V2', $size, $mtime); # 32-bits, may truncate
+	    setfattr($fn, 'rsync.%md4', $szmt.$sum4);
+	    setfattr($fn, 'rsync.%md5', $szmt.$sum5);
+	    #utime $mtime, $mtime, $fn; # Set mtime if it changes.
+	}
+    }
+
+    closedir DP;
+
+    unshift(@dirs, sort @subdirs) if $recurse_opt;
+}
+
+sub usage
+{
+    die <<EOT;
+Usage: rsyncsums [OPTIONS] [DIRS]
+
+Options:
+ -r, --recurse     Update checksums in subdirectories too.
+ -v, --verbose     Mention what we're doing.  Repeat for more info.
+ -h, --help        Display this help message.
+EOT
+}
--- old/xattrs.c
+++ new/xattrs.c
@@ -31,6 +31,8 @@ extern int am_generator;
 extern int read_only;
 extern int list_only;
 extern int checksum_seed;
+extern int checksum_len;
+extern int protocol_version;
 
 #define RSYNC_XAL_INITIAL 5
 #define RSYNC_XAL_LIST_INITIAL 100
@@ -62,7 +64,8 @@ extern int checksum_seed;
 #define RPRE_LEN ((int)sizeof RSYNC_PREFIX - 1)
 
 #define XSTAT_ATTR RSYNC_PREFIX "%stat"
-#define XSTAT_LEN ((int)sizeof XSTAT_ATTR - 1)
+#define MD4_ATTR RSYNC_PREFIX "%md4"
+#define MD5_ATTR RSYNC_PREFIX "%md5"
 
 typedef struct {
 	char *datum, *name;
@@ -223,8 +226,8 @@ static int rsync_xal_get(const char *fna
 			continue;
 #endif
 
-		if (am_root < 0 && name_len == XSTAT_LEN + 1
-		 && name[RPRE_LEN] == '%' && strcmp(name, XSTAT_ATTR) == 0)
+		if (name_len > RPRE_LEN && name[RPRE_LEN] == '%'
+		 && HAS_PREFIX(name, RSYNC_PREFIX))
 			continue;
 
 		datum_len = name_len; /* Pass extra size to get_xattr_data() */
@@ -636,8 +639,8 @@ void receive_xattr(struct file_struct *f
 			continue;
 		}
 #endif
-		if (am_root < 0 && name_len == XSTAT_LEN + 1
-		 && name[RPRE_LEN] == '%' && strcmp(name, XSTAT_ATTR) == 0) {
+		if (name_len > RPRE_LEN && name[RPRE_LEN] == '%'
+		 && HAS_PREFIX(name, RSYNC_PREFIX)) {
 			free(ptr);
 			continue;
 		}
@@ -795,6 +798,39 @@ int set_xattr(const char *fname, const s
 	return rsync_xal_set(fname, lst + ndx, fnamecmp, sxp);
 }
 
+int get_sum_xattr(const char *fname, STRUCT_STAT *stp, char *sum)
+{
+	const char *mdattr = protocol_version >= 30
+			   ? MD5_ATTR : MD4_ATTR;
+	char buf[256];
+	uint32 file_length, mtime;
+	int len;
+	
+	len = sys_lgetxattr(fname, mdattr, buf, sizeof buf);
+	if (len < 0) {
+		if (errno == ENOTSUP || errno == ENOATTR)
+			return 0;
+		rsyserr(FERROR, errno, "failed to read xattr %s for %s",
+			mdattr, full_fname(fname));
+		return 0;
+	}
+	if (len != 4 + 4 + checksum_len) {
+		rprintf(FERROR, "Corrupt %s xattr attached to %s -- skipping\n",
+			mdattr, full_fname(fname));
+		return 0;
+	}
+
+	file_length = IVAL(buf, 0); /* 32-bit values -- trunctions are OK */
+	mtime = IVAL(buf, 4);
+
+	if ((uint32)stp->st_size != file_length || (uint32)stp->st_mtime != mtime)
+		return 0;
+
+	memcpy(sum, buf + 8, checksum_len);
+
+	return 1;
+}
+
 int get_stat_xattr(const char *fname, int fd, STRUCT_STAT *fst, STRUCT_STAT *xst)
 {
 	int mode, rdev_major, rdev_minor, uid, gid, len;
